rig-lancedb/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+        let documents: DocumentRecords = self
+            .document_table
+            .query()
+            .only_if(format!("id = '{id}'"))
+            .execute_query()
+            .await?;
+
+        let embeddings: EmbeddingRecordsBatch = self
+            .embedding_table
+            .query()
+            .only_if(format!("document_id = '{id}'"))
+            .execute_query()
+            .await?;
+
+        Ok(merge(&documents, &embeddings)?.into_iter().next())
+    }
+
+    async fn get_document<T: for<'a> serde::Deserialize<'a>>(
+        &self,
+        id: &str,
+    ) -> Result<Option<T>, VectorStoreError> {
+        let documents: DocumentRecords = self
+            .document_table
+            .query()
+            .only_if(format!("id = '{id}'"))
+            .execute_query()
+            .await?;
+
+        let document = documents
+            .as_iter()
+            .next()
+            .map(|document| serde_json::from_str(&document.document).map_err(serde_to_rig_error))
+            .transpose();
+
+        document
+    }
+
+    async fn get_document_by_query(
+        &self,
+        query: Self::Q,
+    ) -> Result<Option<rig::embeddings::DocumentEmbeddings>, VectorStoreError> {
+        let documents: DocumentRecords = query.execute_query().await?;
+
+        let embeddings: EmbeddingRecordsBatch = self
+            .embedding_table
+            .query()
+            .only_if(format!(
+                "document_id IN ({})",
+                documents
+                    .ids()
+                    .map(|id| format!("'{id}'"))
+                    .collect::<Vec<_>>()
+                    .join(",")
+            ))
+            .execute_query()
+            .await?;
+
+        Ok(merge(&documents, &embeddings)?.into_iter().next())
+    }
+}
+
+impl<M: EmbeddingModel + std::marker::Sync + Send> VectorStoreIndex for LanceDbVectorStore<M> {
+    async fn top_n_from_query(
+        &self,
+        query: &str,
+        n: usize,
+    ) -> Result<Vec<(f64, rig::embeddings::DocumentEmbeddings)>, VectorStoreError> {
+        let prompt_embedding = self.model.embed_document(query).await?;
+        self.top_n_from_embedding(&prompt_embedding, n).await
+    }
+
+    async fn top_n_from_embedding(
+        &self,
+        prompt_embedding: &rig::embeddings::Embedding,
+        n: usize,
+    ) -> Result<Vec<(f64, rig::embeddings::DocumentEmbeddings)>, VectorStoreError> {
+        let query = self
+            .embedding_table
+            .vector_search(prompt_embedding.vec.clone())
+            .map_err(lancedb_to_rig_error)?
+            .limit(n);
+
+        let SearchParams {
+            distance_type,
+            search_type,
+            nprobes,
+            refine_factor,
+            post_filter,
+        } = self.search_params.clone();
+
+        if let Some(distance_type) = distance_type {
+            query.clone().distance_type(distance_type);
```

Comment:
This doesn't actually set anything since the `distance_type` method is invoked on a _clone_ of `query`. 

Make `query` mutable and change all the `query.clone().foo()` to `query = query.foo()`